code/datums/tracking_sound.dm

```diff
@@ -0,0 +1,112 @@
+	/// Set to true if we were able to track sound length for self deletion.
+	var/qdel_scheduled = FALSE
+
+/datum/sound_spatial_tracker/New(source, channel, range, sound_length, playsound_local_args)
+	src.source = source
+	start_time = REALTIMEOFDAY
+	src.channel = channel
+	src.range = range
+	src.playsound_local_args = playsound_local_args
+	GLOB.sound_spatial_trackers["[channel]"] = src
+	spatial_tracker = new(range, range)
+	update_spatial_tracker()
+	if(sound_length)
+		schedule_qdel(sound_length)
+	return ..()
+
+/datum/sound_spatial_tracker/proc/schedule_qdel(length)
+	QDEL_IN(src, length)
+	qdel_scheduled = TRUE
+
+/datum/sound_spatial_tracker/Destroy(force, ...)
+	source = null
+	playsound_local_args.Cut()
+	for(var/mob/listener as anything in listeners)
+		release_listener(listener)
+	listeners.Cut()
+	if(GLOB.sound_spatial_trackers["[channel]"] == src)
+		GLOB.sound_spatial_trackers -= "[channel]"
+	spatial_tracker = null
+	return ..()
+
+/datum/sound_spatial_tracker/proc/on_source_moved()
+	SIGNAL_HANDLER
+	INVOKE_ASYNC(src, PROC_REF(update_spatial_tracker))
+
+/datum/sound_spatial_tracker/proc/update_spatial_tracker()
+	var/list/new_and_old_cells = spatial_tracker.recalculate_cells(get_turf(source))
+	for(var/datum/spatial_grid_cell/new_cell as anything in new_and_old_cells[1])
+		RegisterSignal(new_cell, SPATIAL_GRID_CELL_ENTERED(SPATIAL_GRID_CONTENTS_TYPE_CLIENTS), PROC_REF(entered_cell))
+		for(var/mob/listener as anything in new_cell.client_contents)
+			link_to_listener(listener)
+	for(var/datum/spatial_grid_cell/old_cell as anything in new_and_old_cells[2])
+		UnregisterSignal(old_cell, SPATIAL_GRID_CELL_ENTERED(SPATIAL_GRID_CONTENTS_TYPE_CLIENTS))
+		for(var/mob/listener as anything in old_cell.client_contents)
+			release_listener(listener)
+
+/datum/sound_spatial_tracker/proc/link_to_listener(mob/listener)
+	if(listener in listeners)
+		return
+	RegisterSignal(listener, COMSIG_MOVABLE_MOVED, PROC_REF(on_listener_moved))
+	listeners[listener] = TRUE
+
+/datum/sound_spatial_tracker/proc/release_listener(mob/listener)
+	if(!(listener in listeners))
+		return
+	listeners -= listener
+	UnregisterSignal(listener, COMSIG_MOVABLE_MOVED)
+	if(isnull(listener.client))
+		return
+	var/sound/null_sound = sound(null, channel = channel)
+	SEND_SOUND(listener.client, null_sound)
+
+/datum/sound_spatial_tracker/proc/update_listener(mob/listener)
+	var/sound/existing_sound = null
+	for(var/sound/playing as anything in listener.client?.SoundQuery())
+		if(playing.channel != channel)
+			continue
+		existing_sound = playing
+
+	var/expected_offset = (REALTIMEOFDAY - start_time) * 0.1
+	if(!qdel_scheduled && !isnull(existing_sound)) // couldn't do it before, but we can now
+		var/actual_len = existing_sound.len * 10 // WHY IS THIS IN SECONDS
+		schedule_qdel(actual_len - expected_offset)
+
+	var/list/new_args = playsound_local_args.Copy()
+	new_args["turf_source"] = get_turf(source)
+
+	var/sound/new_sound = sound(new_args["soundin"])
```

Comment:
then just update the sound datum's offset here instead of making a new one each time